lib/index.js

```diff
@@ -121,99 +124,90 @@ Optimizely.prototype.page = function(page) {
-    // this value is made available
-    if (experiment.referrer) props.referrer = experiment.referrer;
+    // this value is made available. So if a customer came in via google.com/ad -> tb12.com -> redirect experiment -> Belichickgoat.com
+    // `experiment.referrer` would be google.com/ad here NOT `tb12.com`.
+    if (experiment.referrer) {
+      props.referrer = experiment.referrer;
+      context.page = { referrer: experiment.referrer };
+    }
 
     // When there is a multivariate experiment
     if (sections) {
       // Since `sections` include all the possible sections on the page, we need to find the names of the sections
       // if any of its variations were used. Experiments could display variations from multiple sections.
-      // So we have to do these terrible for loops as a workaround.
-      var sectionNames = [];
-      var sectionIds = [];
-      var sectionMap = {};
+      // The global optimizely data object does not expose a mapping between which section(s) were involved within an experiment.
+      // So we will build our own mapping to quickly get the section name(s) and id(s) for any displayed variation.
       var activeSections = {};
-
-      // We need to reformat the data structure to reduce complexity from 4x to 2x
-      for (var sectionId in sections) {
-        if (!sections.hasOwnProperty(sectionId)) continue;
-        var section = sections[sectionId];
-        var sectionVariationIds = section.variation_ids;
-        for (var i = 0; i < sectionVariationIds.length; i++) {
-          sectionMap[sectionVariationIds[i]] = {
-            id: sectionId,
-            name: section.name
-          };
-        }
-      }
-
-      // Experiments can display variations from multiple sections so we are checking which
-      // sections were actually used
+      var variationIdsToSectionsMap = foldl(function(results, section, sectionId) {
+        each(function(variationId) {
+          results[variationId] = { id: sectionId, name: section.name };
+        }, section.variation_ids);
+        return results;
+      }, {}, sections);
       for (var j = 0; j < variationIds.length; j++) {
-        var activeSection = sectionMap[variationIds[j]];
-        if (activeSection && !activeSections[activeSection.id]) {
-          sectionNames.push(activeSection.name);
-          sectionIds.push(activeSection.id);
-          activeSections[activeSection.id] = true; // For deduping purposes
-        }
+        var activeVariation = variationIds[j];
+        var activeSection = variationIdsToSectionsMap[activeVariation];
+        if (activeSection) activeSections[activeSection.id] = activeSection.name;
       }
 
-      props.sectionName = sectionNames.join(', ');
-      props.sectionId = sectionIds.join();
+      props.sectionId = keys(activeSections).join(); // Not adding space for backward compat/consistency reasons since all IDs we've never had spaces
```

Comment:
These could be in a different order depending on the browser. We should probably sort the ids and names so there is a consistent value for the same active sections.